enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -229,35 +230,119 @@ doing all of your testing in a virtual network built on top of it.
+will be to enhance MetalLB with a native CRD based configuration interface.
+The design work has begun here:
+
+https://github.com/metallb/metallb/pull/833
+
+Since this is an upstream effort, the completion of this work may take longer
+than the timeline for when we want to ship the first piece of MetalLB support
+for OpenShift, which will be to support MetalLB in layer2 mode only.
+
+To mitigate this timing risk, we have another option for a way to configure the
+layer2 mode of MetalLB by adding a small new feature to OpenShift's existing
+network config object.  OpenShift has an existing feature to automatically
+assign external IPs to services of `type=LoadBalancer`.  For example, take this
+sample configuration:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Network
+metadata:
+  name: cluster
+spec:
+  ...
+  externalIP:
+    autoAssignCIDRs:
+    - 10.10.10.0/24
+```
+
+This makes OpenShift automatically set the `externalIP` field, but nothing will
+make sure those addresses are reachable automatically.  What we need is a
+similar option that instead of setting `ExternalIPs`, it should set the
+`LoadBalancerIP` field in the Service spec.  The proposed new field is
+`loadBalancer.autoAssignCIDRs`.  With the new option, the way to configure the
+address pool for MetalLB would be:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Network
+metadata:
+  name: cluster
+spec:
+  ...
+  loadBalancer:
+    autoAssignCIDRs:
+    - 10.10.10.0/24
+```
+
+The openshift-controller-manager should have new code similar to its handling
+of `externalIP.autoAssignCIDRs` for implementing the Services address
+allocation for this new option.
+
+At this point we will have OpenShift automatically allocating IP addresses to
+Services of type LoadBalancer. MetalLB will be configured to honor this
+requested IP address. If the OpenShift MetalLB operator is enabled (the
+operator has been installed via OLM), the operator can automatically generate a
+MetalLB config map based on this cluster configuration.  The operator will read
+the cluster network configuration and generate config like this:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: metallb-system
+  name: config
+data:
+  config: |
+    address-pools:
+    - name: openshift
+      protocol: layer2
+      auto-assign: false
+      addresses:
+      - 10.10.10.0/24
+```
+
+Note how the configuration copies over the CIDR for the LoadBalancerIP allocator
+used in OpenShift.  This makes MetalLB expect to see addresses in this range
+used for load balancer Services.  Also note the use of `auto-assign: false`.
+In this scenario, we're letting OpenShift do the address allocation piece.
+
+If we support this as a configuration interface, customers will be able to opt
+for the more feature rich CRD interface offered by MetalLB natively
+once it is ready.  As long as the `loadBalancer.autoAssignCIDRs` feature is not
+enabled on the OpenShift side, there should not be a conflict.
```

Comment:
so wait, does MetalLB already support `auto-assign: true`? If so, then why are we reimplementing it in OCP?

Reply:
It does, yes, but via its ConfigMap config interface, and we want to only direct customers to use CRDs.  We're working on a full CRD config interface upstream, but it's unlikely to be ready in time for 4.9.  There's a link to that inline.

This is one approach to having an OpenShift specific CRD interface: a generally useful API for configuring load balancer IPAM (if it is indeed useful for any other platform, not actually sure).